luda-editor/client/src/pages/graphic_asset_manage_page/cg_viewer/part_picker.rs

```diff
@@ -0,0 +1,474 @@
+use crate::{color, components::tool_tip::ToolTip, storage::get_project_cg_part_variant_image_url};
+use namui::prelude::*;
+use namui_prebuilt::{
+    scroll_view, simple_rect,
+    table::hooks::*,
+    typography::{center_text, center_text_full_height},
+};
+use rpc::data::{CgFile, CgPart, CgPartVariant, PartSelectionType, ScreenCg};
+use std::{iter::once, ops::Deref};
+
+const THUMBNAIL_WH: Wh<Px> = Wh {
+    width: px(96.0),
+    height: px(96.0),
+};
+const MIN_THUMBNAIL_CONTAINER_SIDE_PADDING: Px = px(8.0);
+const ROW_VERTICAL_PADDING: Px = px(8.0);
+const DIVIDER_HEIGHT: Px = px(32.0);
+
+#[namui::component]
+pub struct PartPicker<'a> {
+    pub wh: Wh<Px>,
+    pub cg_file: &'a CgFile,
+    pub project_id: Uuid,
+    pub screen_cg: &'a ScreenCg,
+    pub on_event: &'a dyn Fn(Event),
+}
+
+pub enum Event {
+    UnselectCgPart {
+        cg_part_name: String,
+    },
+    TurnOnCgPartVariant {
+        cg_part_name: String,
+        cg_part_variant_name: String,
+    },
+    TurnOffCgPartVariant {
+        cg_part_name: String,
+        cg_part_variant_name: String,
+    },
+}
+
+pub enum InternalEvent {
+    MouseHoverInThumbnail { global_xy: Xy<Px>, name: String },
+    MouseHoverOutThumbnail,
+}
+
+impl Component for PartPicker<'_> {
+    fn render(self, ctx: &RenderCtx) -> RenderDone {
+        const PADDING: Px = px(8.0);
+
+        let Self {
+            wh,
+            cg_file,
+            project_id,
+            screen_cg,
+            on_event,
+        } = self;
+
+        let (mouse_hovering_part_variant, set_mouse_hovering_part_variant) =
+            ctx.state::<Option<MouseHoveringPartVariant>>(|| None);
+
+        let on_internal_event = |internal_event| match internal_event {
+            InternalEvent::MouseHoverInThumbnail { global_xy, name } => {
+                set_mouse_hovering_part_variant
+                    .set(Some(MouseHoveringPartVariant { name, global_xy }))
+            }
+            InternalEvent::MouseHoverOutThumbnail => set_mouse_hovering_part_variant.set(None),
+        };
```

Comment:
You don't have to do this. You can just put logic inside of compose code, just like react.
If you need to make a variable, that's only when your code is too long, then just make a variable, not enum, just like React.

Reply:
I don't have to do this. but Isn't it better to drill event handler to descendant component?

Reply:
@namse Did you see this comment?

Reply:
It should be descendant's event handler then, not the `internal` one.

Reply:
Ah, that must be confusing.
I think that it is internal_event because this event is used only in the same file, `part_picker.rs`.
But you think internal_event should only be used in one component, right?

Reply:
Personally I think the term of `internal` is obsolete because we can just call closure in single component without wrapping internal event enum.

Reply:
I agree with that.

What do you think of using `internal_event` for multiple components in single file, not a single component? like this case.
It is very handy way.

Reply:
@namse If it does make sense, please resolve conversation please.

Reply:
i have no idea about that. use it when you want to use it and when it makes sense.